src/proxy/ssh/server.ts

```diff
@@ -0,0 +1,1170 @@
+      return `${bytes} bytes`;
+    }
+
+    const units = ['bytes', 'KB', 'MB', 'GB', 'TB'];
+    let value = bytes;
+    let unitIndex = 0;
+
+    while (value >= 1024 && unitIndex < units.length - 1) {
+      value /= 1024;
+      unitIndex++;
+    }
+
+    const precision = unitIndex === 0 ? 0 : 2;
+    return `${value.toFixed(precision)} ${units[unitIndex]}`;
+  }
+
+  async handleClient(
+    client: ssh2.Connection,
+    clientInfo?: { ip?: string; family?: string },
+  ): Promise<void> {
+    const clientIp = clientInfo?.ip || 'unknown';
+    console.log(`[SSH] Client connected from ${clientIp}`);
+    const clientWithUser = client as ClientWithUser;
+    clientWithUser.clientIp = clientIp;
+
+    // Set up connection timeout (10 minutes)
+    const connectionTimeout = setTimeout(() => {
+      console.log(`[SSH] Connection timeout for ${clientIp} - closing`);
+      client.end();
+    }, 600000); // 10 minute timeout
+
+    // Set up client error handling
+    client.on('error', (err: Error) => {
+      console.error(`[SSH] Client error from ${clientIp}:`, err);
+      clearTimeout(connectionTimeout);
+      // Don't end the connection on error, let it try to recover
+    });
+
+    // Handle client end
+    client.on('end', () => {
+      console.log(`[SSH] Client disconnected from ${clientIp}`);
+      clearTimeout(connectionTimeout);
+      // Clean up keepalive timer
+      const keepaliveTimer = this.keepaliveTimers.get(client);
+      if (keepaliveTimer) {
+        clearInterval(keepaliveTimer);
+        this.keepaliveTimers.delete(client);
+      }
+    });
+
+    // Handle client close
+    client.on('close', () => {
+      console.log(`[SSH] Client connection closed from ${clientIp}`);
+      clearTimeout(connectionTimeout);
+      // Clean up keepalive timer
+      const keepaliveTimer = this.keepaliveTimers.get(client);
+      if (keepaliveTimer) {
+        clearInterval(keepaliveTimer);
+        this.keepaliveTimers.delete(client);
+      }
+    });
+
+    // Handle keepalive requests
+    (client as any).on('global request', (accept: () => void, reject: () => void, info: any) => {
+      console.log('[SSH] Global request:', info);
+      if (info.type === 'keepalive@openssh.com') {
+        console.log('[SSH] Accepting keepalive request');
+        // Always accept keepalive requests to prevent connection drops
+        accept();
+      } else {
+        console.log('[SSH] Rejecting unknown global request:', info.type);
+        reject();
+      }
+    });
+
+    // Handle authentication
+    client.on('authentication', (ctx: ssh2.AuthContext) => {
+      console.log(
+        `[SSH] Authentication attempt from ${clientIp}:`,
+        ctx.method,
+        'for user:',
+        ctx.username,
+      );
+
+      if (ctx.method === 'publickey') {
+        // Handle public key authentication
+        const keyString = `${ctx.key.algo} ${ctx.key.data.toString('base64')}`;
+
+        (db as any)
```

Comment:
Is casting necessary here? I thought that `db` is exporting `findUserBySSHKey` 🤔